libbitstream/bitstream.c

```diff
@@ -0,0 +1,313 @@
+// * Neither the name  of Intel Corporation  nor the names of its contributors
+//   may be used to  endorse or promote  products derived  from this  software
+//   without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING,  BUT NOT LIMITED TO,  THE
+// IMPLIED WARRANTIES OF  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED.  IN NO EVENT  SHALL THE COPYRIGHT OWNER  OR CONTRIBUTORS BE
+// LIABLE  FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR
+// CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
+// SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
+// INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
+// CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif // HAVE_CONFIG_H
+
+#include <errno.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include <uuid/uuid.h>
+#include <json-c/json.h>
+#include "bitstream.h"
+
+#include "safe_string/safe_string.h"
+#include <opae/log.h>
+#include <opae/properties.h>
+#include <opae/sysobject.h>
+
+#define METADATA_GUID     "58656F6E-4650-4741-B747-425376303031"
+#define METADATA_GUID_LEN 16
+#define GBS_AFU_IMAGE     "afu-image"
+#define BBS_INTERFACE_ID  "interface-uuid"
+
+/*
+ * Check for bitstream header and fill out bistream_info fields
+ */
+#define MAGIC       0x1d1f8680
+#define MAGIC_SIZE  4
+#define HEADER_SIZE 20
+STATIC int parse_metadata(opae_bitstream_info *info)
+{
+	unsigned i;
+
+	if (!info)
+		return -EINVAL;
+
+	if (info->data_len < HEADER_SIZE) {
+		OPAE_ERR("File too small to be GBS.");
+		return -1;
+	}
+
+	if (((uint32_t *)info->data)[0] != MAGIC) {
+		OPAE_ERR("No valid GBS header.");
+		return -1;
+	}
+
+	/* reverse byte order when reading GBS */
+	for (i = 0; i < sizeof(info->interface_id); i++)
+		info->interface_id[i] =
+			info->data[MAGIC_SIZE+sizeof(info->interface_id)-1-i];
+
+	info->rbf_data = &info->data[HEADER_SIZE];
+	info->rbf_len = info->data_len - HEADER_SIZE;
+
+	return 0;
+}
+
+STATIC fpga_result string_to_guid(const char *guid, fpga_guid *result)
+{
+	if (uuid_parse(guid, *result) < 0) {
+		OPAE_ERR("Error parsing guid: %s.", guid);
+		return FPGA_INVALID_PARAM;
+	}
+
+	return FPGA_OK;
+}
+
+STATIC fpga_result check_bitstream_guid(const uint8_t *bitstream)
+{
+	fpga_guid bitstream_guid;
+	fpga_guid expected_guid;
+	errno_t e;
+
+	e = memcpy_s(bitstream_guid, sizeof(bitstream_guid), bitstream,
+		     sizeof(fpga_guid));
+	if (EOK != e) {
+		OPAE_ERR("memcpy_s failed.");
+		return FPGA_EXCEPTION;
+	}
+
+	if (string_to_guid(METADATA_GUID, &expected_guid) != FPGA_OK)
```

Comment:
Would it make sense to encode the metadata GUID as raw bytes?

Reply:
Will note as a TODO for @anandaravuri to evaluate.